Automatic Authentication for Schema Registry and HTTP Proxy - part1

BenPope opened the pull request:
## Cover letter

Implement a mechanism to share ephemeral credentials over internal RPC to allow Schema Registry and Pandaproxy to obtain credentials for a sasl enabled endpoint, without explicitly specifying the secret in the configuration file.

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/1209c39e567090535f898082e66575d86610353f..c0b69c561e2e9e3cbabd370951ea1a3c551948b7)

* Reimplementation to avoid a new auth type
* Rebase due to several conflicts

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c0b69c561e2e9e3cbabd370951ea1a3c551948b7..46e939cbfec7f43c4a0086d02775241a63d85304)
* Rebase over conflict
* Improve error handling in frontend by
  * Moving to `std::error_code`
  * Properly capturing rpc errors
* Fix trailing newlines
* Improve some logging
* Bump the serde version of `scram_credential`

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/46e939cbfec7f43c4a0086d02775241a63d85304..9a89cbd60cd816dda0ab87cfecfea6360d82f7a7)
* Rebase over conflict (again)

Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9a89cbd60cd816dda0ab87cfecfea6360d82f7a7..3cb3969eae7610313b3ad76f1ba43e73a73a6654)
* Initialise `ephemeral_credential_frontend::_self` (fixes CI failure)
* `ephemeral_credential_frontend::get` now returns just the `ephemeral_credential`
* Don't serialise `scram_credential::principal`, don't bump serde::version
* Make `find` and `has` members of `ephemeral_credential_store` `const`
* Don't serialize `ephemeral_credential`
* Remove `entire_credential`
* Have the test rely on fewer implementation details

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [x] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

* none

## Release notes

* none

dotnwat reviewed (commented):
generally looks like a good direction.

but i'm not sure i fully understand the need for a new scram_credential type in so much as it seems to be causing a lot of fall out in terms of new code (e.g. new store) and more generated code (e.g. scram_authenticator).

would a similar effect be achieved by adding extra state to the existing scram* objects or is the type serving a special purpose i'm not seeing?

dotnwat commented on src/v/security/scram_authenticator.cc:
```diff
@@ -16,21 +16,20 @@
     _authid = _client_first->username_normalized();
-    auto credential = _credentials.get<scram_credential>(
-      credential_user(_authid));
+    auto credential = _credentials.get<P>(credential_user(_authid));
```
i think this comment is mostly superseded by the  top level review comment

BenPope replied:
This has been reworked - the mapped `principal` is now stored in the `scram_credential`.

NyaliaLui commented on src/v/features/feature_table.h:
```diff
@@ -181,6 +182,12 @@ constexpr static std::array feature_schema{
     feature_spec::available_policy::always,
     feature_spec::prepare_policy::always},
+  feature_spec{
+    cluster_version{7},
```
Not a comment for reviews but more for my understanding of cluster version. From the code base it seems `cluster version` indicates the type of content on raft0. My question is: What content does version number 7 refer to?

BenPope replied:
It's a number that represents the feature set that's supported. We bump it when we've added new features, and make a release. (Well, we bump it at the first new feature on dev).

dotnwat replied:
@NyaliaLui i won't leave the link here, but there is an internal wiki page on the feature that you can find in confluence.

NyaliaLui commented on src/v/security/ephemeral_credential.h:
```diff
@@ -0,0 +1,85 @@
+    friend std::ostream& operator<<(std::ostream&, const ephemeral_credential&);
+
+    acl_principal _principal;
+    credential_user _user;
```
nit: Point of clarification, why have a credential_user when the acl_principal typically also contains the username?

BenPope replied:
It's expected in the future that there will be additional `principal_types` such as `group`. This PR somewhat hijacks that mechanism to create an `ephemeral_user` type.

BenPope replied:
This PR allows some decoupling of the `credential_user` from the `acl_principal`. Now there can be multiple `ephemeral_credentials`, each with a different `user` and `password`, that are associated with the `acl_principal`.

NyaliaLui commented on src/v/cluster/ephemeral_credential_service_impl.json:
```diff
@@ -0,0 +1,14 @@
+            "output_type": "put_ephemeral_credential_reply"
+        }
+    ]
+}
```
nit: newline

Or do we not do that for json files?

BenPope replied:
Done

NyaliaLui commented on src/v/cluster/ephemeral_credential_service.h:
```diff
@@ -0,0 +1,32 @@
+    ss::sharded<ephemeral_credential_frontend>& _fe;
+};
+
+} // namespace cluster
```
nit: newline

BenPope replied:
Done

NyaliaLui reviewed (commented):
Is there an RFC for auto auth or maybe some other documentation? It would help me with the reviews to understand the need for ephemeral_credential and ephemeral_credential_store.
I simply cannot tell if the pre-existing credential and store objects would suffice.

Feel free to ping me on slack.

The code itself looks good, I just need some clarification.

NyaliaLui commented:
> Is there an RFC for auto auth or maybe some other documentation? It would help me with the reviews to understand the need for ephemeral_credential and ephemeral_credential_store. I simply cannot tell if the pre-existing credential and store objects would suffice.

Oh I found the RFC but it looks like it may still be a draft? Regardless I would still like to know the design decisions behind creating a new credential object and store.

BenPope commented on src/v/security/scram_credential.h:
```diff
@@ -47,6 +62,7 @@ class scram_credential
     bytes _server_key;
     bytes _stored_key;
     int _iterations{0};
+    std::optional<acl_principal> _principal;
```
I need to fix the versioned serialisation for this.

BenPope replied:
Is this sufficient?
```
-  : public serde::envelope<scram_credential, serde::version<0>> {
+  : public serde::
+      envelope<scram_credential, serde::version<1>, serde::compat_version<0>> {
```

dotnwat replied:
yeh looks good

BenPope replied:
I've avoided bumping the version by not serialising the principal.

BenPope commented on src/v/cluster/tests/ephemeral_credential_test.cc:
```diff
@@ -0,0 +1,104 @@
+
+    // Check inform works
+    {
+        auto res = fe_0.inform(apps[1]->controller->self(), principal).get();
```
The CI failure here is due to `rpc - connection_cache.h:90 - No client available`, is there something else I should wait on?

BenPope replied:
`_self` was not initialised. Fixed.

dotnwat reviewed (commented):
im generally onboard with this approach.

a few minor things that seem to be unrelated to the overall architecture:

* i don't understand the relationship between the three credential types and why they exist
* in at least one of them i see a password is serialized. it's not clear why we would ever need to serialize a password when using scram mechanism
* it doesn't seem like the scram_credential needs to serialize a principal. seems like it might be convenient to include it in the in-memory representation but not clear why it goes over the wire

to the last point, maybe the principal could be in the rpc message and then added to the in-memory representation when the credential is added to the store?

BenPope commented on src/v/security/ephemeral_credential.h:
```diff
@@ -0,0 +1,85 @@
+namespace security {
+
+class ephemeral_credential
+  : public serde::envelope<ephemeral_credential, serde::version<0>> {
```
It looks like most types that inherit from `envelope` are a `struct`, so they're publicly inherited.

BenPope replied:
This is no longer serialisable.

BenPope commented on src/v/security/ephemeral_credential.cc:
```diff
@@ -0,0 +1,27 @@
+namespace security {
+
+std::ostream& operator<<(std::ostream& os, const ephemeral_credential& c) {
+    fmt::print(os, "principal: {}, user: {}", c.principal(), c.user());
```
It's pretty hardcoded to 512 for ephemeral credentials, I mostly wanted to log the mapping between user and principal for debugging.

BenPope commented on src/v/security/ephemeral_credential_store.h:
```diff
@@ -0,0 +1,87 @@
+    }
+
+private:
+    static void assert_principal(acl_principal const& p) {
```
I guess acl_principal could be a variant over the `principal_type`, but it would probably get pretty awkward. This check is mostly to spot incorrect logic - there should be no way to get `principal_type::user` into this store.

BenPope commented on src/v/security/ephemeral_credential_store.h:
```diff
@@ -0,0 +1,87 @@
+        }
+        auto n = _credentials.extract(it);
+        n.value() = std::move(s);
+        return _credentials.insert(std::move(n)).position;
```
Nothing. For the node operations added in c++17, they decided to use a struct with named parameters, rather than a `std::pair<iterator, bool>`, so `::insert(node_handle)` returns the struct.

BenPope commented on src/v/security/types.h:
```diff
@@ -0,0 +1,25 @@
+namespace security {
+
+using credential_user = named_type<ss::sstring, struct credential_user_type>;
+using credential_password
```
We have something similar for config, perhaps that mechanism could be generalised, but `operator<<` is pretty manual to implement anyway.

I don't really like `operator<<` or `toString` general, converting a type to a string is contextual, there's unstructured logging for humans, `yaml`, `json`, `xml`, etc, and even those can be contextual.

BenPope commented on src/v/security/ephemeral_credential_store.h:
```diff
@@ -0,0 +1,87 @@
+public:
+    ephemeral_credential_store() = default;
+
+    const_iterator find(acl_principal const& p) {
```
Done

BenPope commented:
> im generally onboard with this approach.
> 
> a few minor things that seem to be unrelated to the overall architecture:
> 
> * i don't understand the relationship between the three credential types and why they exist

I removed `entire_credential` and added a doc for `ephemeral_credential`.

> * in at least one of them i see a password is serialized. it's not clear why we would ever need to serialize a password when using scram mechanism

Thank you. This is no longer serialised.

> * it doesn't seem like the scram_credential needs to serialize a principal. seems like it might be convenient to include it in the in-memory representation but not clear why it goes over the wire

Thanks again, done.

> to the last point, maybe the principal could be in the rpc message and then added to the in-memory representation when the credential is added to the store?

Done.

dotnwat commented on src/v/cluster/ephemeral_credential_frontend.cc:
```diff
@@ -0,0 +1,179 @@
+        res.credential = make_ephemeral_credential(principal);
+        co_await put(res.credential);
+    } else {
+        res.credential = *it;
```
i never would have guessed that `it` was in scope here!

BenPope replied:
Yeah, it's a bit unexpected, but it's not like the scope is meaningfully different than the true case.

dotnwat commented on src/v/cluster/ephemeral_credential_frontend.cc:
```diff
@@ -0,0 +1,179 @@
+        res.credential = *it;
+    }
+
+    if (!_c_store.local().contains(res.credential.user())) {
+        co_await put(
+          res.credential.principal(),
+          res.credential.user(),
+          make_scram_credential(res.credential));
+    }
```
it's not immediately obvious why the credential needs to go into the regular credential store here. presumably this code path is for the client, so making and storing an ephemeral credential makes sense. but then, adding to the normal store, would that imply that some broker is going to try to authenticate against _this_ broker that the client is on with the same credential?

BenPope replied:
I've deleted and added this code back several times!

Currently It's so that it can be looked up for the inform call. But that's kinda nonsense, because it can be generated and there's no need to store it locally.

BenPope replied:
OK, this breaks a test. Let me change the test to be what it should have tested in the first place; that the other broker can authenticate the ephemeral_credentials.

dotnwat replied:
cool. other than resolving this question this pr is pretty much ready to go

piyushredpanda replied:
Yay!

BenPope replied:
It's a bit too complicated to test because I'd have to go through the whole scram dance due to not having the credentials that are sent, and the randomness of the salt.

BenPope replied:
Actually, not that hard.

BenPope commented on src/v/cluster/ephemeral_credential_frontend.cc:
```diff
@@ -0,0 +1,179 @@
+    };
+    return {security::ephemeral_credential{
+      principal,
+      security::credential_user{gen()},
```
That might be useful. In phase 2, I was going to update the [configuration with a client id](https://github.com/redpanda-data/redpanda/pull/6585/commits/551d9a82b98d43343618fa0e7c49011bd6c68f7b), but I'm not sure if that's logged anywhere.

Also, we talked about a UUID, maybe I should fix that, too.

BenPope commented on src/v/security/scram_authenticator.cc:
```diff
@@ -24,17 +24,18 @@ scram_authenticator<T>::handle_client_first(bytes_view auth_bytes) {
     if (!credential) {
         return errc::invalid_credentials;
     }
+    _principal = credential->principal().value_or(
+      acl_principal{principal_type::user, authid});
```
Exactly. Previously "building a normal user" was handled in `bool connection_context::authorized_user`.

BenPope commented:
Changes in [force-pushes](https://github.com/redpanda-data/redpanda/compare/8509320586ce818ee65e3e3ec93450a3374aff14..60fefbf2f81baafa4723de592dc5dfe77978bf99)
* Don't store the `scram_credential` locally.
* Test that broker 1 has credentials that can be authenticated with the ephemeral credential broker 1 has.

The trade-off is that every `inform` creates a new credential, and they're not cleaned up.

dotnwat commented:
Changes in [force-pushes](https://github.com/redpanda-data/redpanda/compare/8509320586ce818ee65e3e3ec93450a3374aff14..60fefbf2f81baafa4723de592dc5dfe77978bf99)

Don't store the scram_credential locally.
Test that broker 1 has credentials that can be authenticated with the ephemeral credential broker 0 has.
The trade-off is that every inform creates a new scram_credential on the remote, and they're not cleaned up.

Actually, the same username and password are used, but the salt is random, so the scram_credential on the informed is overwritten.

BenPope commented:
Raised a ticket for the unrelated CI Failure: https://github.com/redpanda-data/redpanda/issues/6889

dotnwat commented:
Looks like https://github.com/redpanda-data/redpanda/issues/6745 may have come back

https://buildkite.com/redpanda/redpanda/builds/17088#018400e2-feed-4ab3-9d90-9e815bb62bd9/6-1284

dotnwat commented:
The other failure is https://github.com/redpanda-data/redpanda/issues/6892 but I really don't think we should block this PR because of it. It was a timeout that rpk encountered when a broker took too long to startup. The broker did appear to startup a few seconds after rpk timed out. I filed the ticket with details. Perhaps a timeout needs to be tuned. It's also slow debug build.
